### PR TITLE
sam: fix CIGAR sequence length validation

### PR DESCRIPTION
**Pull request by brentp**

### CODE REVIEW

**Comment by kortschak** [comment]
I'm not convinced this is the right fix, but if a zero length `Cigar` *is* valid for all sequence lengths then that should be documented and the early return should be done:

```
if len(c) == 0 {
    return true
}
```

However, an alternative is to check in [`Record.UnmarshalSAM`](https://github.com/biogo/hts/blob/master/sam/record.go#L258).

```
	if !bytes.Equal(f[9], []byte{'*'}) {
		r.Seq = NewSeq(f[9])
		if len(r.Cigar) != 0 && !r.Cigar.IsValid(r.Seq.Length) {
			return errors.New("sam: sequence/CIGAR length mismatch")
		}
	}
```

This better matches the semantics in the spec, that the information is not available so don't check, rather than an absent value means everything is valid.

(commit message should be "sam: fix CIGAR sequence length validation" - pref with an issue)

**Comment by brentp** [comment]
done.

File: sam/record.go [hunk 1]
```diff
@@ -255,7 +255,7 @@ func (r *Record) UnmarshalSAM(h *Header, b []byte) error {
 	}
 	if !bytes.Equal(f[9], []byte{'*'}) {
 		r.Seq = NewSeq(f[9])
-		if !r.Cigar.IsValid(r.Seq.Length) {
+		if !(len(r.Cigar) == 0 || r.Cigar.IsValid(r.Seq.Length)) {
```
Comment (kortschak) [inline sam/record.go]: `len(r.Cigar) != 0 && !r.Cigar.IsValid(r.Seq.Length)`.

Comment (brentp) [inline sam/record.go]: why is that better? I prefer the way I wrote it?

Comment (kortschak) [inline sam/record.go]: When I read the negation distributed over the conjunction I need to understand both parts of the conjunction and then negate. With the disjunction I can understand as I read.

Comment (brentp) [inline sam/record.go]: noted

Comment (brentp) [inline sam/record.go]: and changed

**Pull request merged**